demo/Alsa cpp/main.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+ * 
+ * Project: Porcupine Alsa
+ * Created Date: Tuesday, September 18th 2018, 7:41:55 pm
+ * Author: Greg
+ * -----
+ * Last Modified: Tue Sep 18 2018
+ * Modified By: Greg
+ * -----
+ * Copyright (c) 2018 Greg
+ * 
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
+ * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ * 
+ */
+
+
+#include <iostream>
+#include <alsa/asoundlib.h>
+#include "../../include/pv_porcupine.h"
+
+
+using namespace std;
+
+#define NO_ERROR	0
+#define OPEN_ERROR	1
+#define MALLOC_ERROR	2
+#define ANY_ERROR	3
+#define ACCESS_ERROR	4
+#define FORMAT_ERROR	5
+#define RATE_ERROR	6
+#define CHANNELS_ERROR	7
+#define PARAMS_ERROR	8
+#define PREPARE_ERROR	9
+#define SNDREAD_ERROR	10
+#define START_ERROR	11
+#define START_PORCUPINE_ERROR	12
+
+#define MAX_BUF_SIZE	512
+#define MAX_SAMPLES	256000
+
+// Change these to match your environment!
+static char MODEL_PATH[] = "../../lib/common/porcupine_params.pv";
+static char KEYWORD_PATH[] = "../../resources/keyword_files/pineapple_linux.ppn";
+
+static pv_porcupine_object_t *porcupineObject;
+
+static char snd_device[] = "default";
+snd_pcm_t* capture_handle;
+snd_pcm_hw_params_t* hw_params;
+snd_pcm_info_t* s_info;
+unsigned int srate = 16000;
+unsigned int nchan = 1;
+
+
+int init_soundcard(const char *devicename)
+{
+    //connect to and set parameters for the sound device
+    //returns any error setting or connecting
+
+	int err = 0;
+
+    if (devicename)
+        strcpy(snd_device, devicename);
```

Comment:
this is tricky! Because if the devicename is longer than memory allocated to snd_device you will overwrite memory! Do we need to copy ?

Reply:
think I will just make it "default" as a const.. that way the ./asoundrc or the /etc/asound.conf will be used

Reply:
OK